contracts/orderbook/Logic/OrderBook.sol

```diff
@@ -0,0 +1,387 @@
+        bytes32 loanID,
+        uint256 amount,
+        bytes memory loanDataBytes
+    ) internal {
+        address(protocol).call(
+            abi.encodeWithSelector(
+                protocol.closeWithSwap.selector,
+                loanID,
+                trader,
+                amount,
+                false,
+                loanDataBytes
+            )
+        );
+    }
+
+    function getFeed() public view returns (address) {
+        return protocol.priceFeeds();
+    }
+
+    function _isActiveLoan(bytes32 ID) internal view returns (bool) {
+        return protocol.loans(ID).active;
+    }
+
+    function clearOrder(bytes32 orderID) public view returns (bool) {
+        IOrderBook.Order memory order = _allOrders[orderID];
+        if (order.timeTillExpiration < block.timestamp) {
+            return true;
+        }
+        uint256 amountUsed = order.collateralTokenAmount +
+            order.loanTokenAmount;
+        uint256 swapRate;
+        if (order.orderType == IOrderBook.OrderType.LIMIT_OPEN) {
+            swapRate = queryRateReturn(
+                order.loanTokenAddress,
+                order.base,
+                amountUsed
+            );
+        } else {
+            swapRate = queryRateReturn(
+                order.base,
+                order.loanTokenAddress,
+                amountUsed
+            );
+        }
+        if (
+            (
+                order.amountReceived > swapRate
+                    ? (order.amountReceived - swapRate) >
+                        (order.amountReceived * 25) / 100
+                    : (swapRate - order.amountReceived) > (swapRate * 25) / 100
+            )
+        ) {
+            return true;
+        }
+        return false;
+    }
+
+    function getClearOrderList(uint start, uint end) external view returns (bool hasOrders, bytes memory payload) {
+        require(end<=_allOrderIDs.length(), "OrderBook: end is past max orders");
+        bytes32[] memory fullList = new bytes32[](7);
+        uint iter = 0;
+        bytes32 ID;
+        for (uint256 i = start; (i < end && iter < 7);) {
+            ID = _allOrderIDs.at(i);
+            if (clearOrder(ID)) {
+                hasOrders = true;
+                fullList[iter] = ID;
+                ++iter;
+            }
+            unchecked { ++i; }
+        }
+        payload = abi.encode(fullList);
+    }
+
+    function getExecuteOrder(uint start, uint end) external returns (bytes32 ID) {
+        require(end<=_allOrderIDs.length(), "OrderBook: end is past max orders");
+        bytes32 tempID;
+        for (uint256 i = start; i < end;) {
+            tempID = _allOrderIDs.at(i);
+            if (prelimCheck(tempID)) {
+                return tempID;
+            }
+            unchecked { ++i; } 
+        }
+    }
+
+    function queryRateReturn(
+        address start,
+        address end,
+        uint256 amount
+    ) public view returns (uint256) {
+        return IPriceFeeds(getFeed())
```

Comment:
we better store getFeed() locally. this will save one call on every queryRateReturn function call

Reply:
Yes, would save a bit on gas.